packages/next/src/server/app-render/fast-set-immediate.external.ts

```diff
@@ -0,0 +1,531 @@
+      signal.addEventListener(
+        'abort',
+        () => {
+          patchedClearImmediate(immediate)
+          reject(signal.reason)
+        },
+        { once: true }
+      )
+    }
+  })
+}
+
+patchedSetImmediate[promisify.custom] = patchedSetImmediatePromise
+
+const patchedClearImmediate = (
+  immediateObject: NodeJS.Immediate | undefined
+) => {
+  // NOTE: we defensively check for patched immediates even if we're not
+  // currently capturing immediates, because the objects returned from
+  // the patched setImmediate can be kept around for arbitrarily long.
+  // As an optimization, we only do this if the patch was enabled at least once --
+  // otherwise, no patched objects could've been created.
+  if (
+    wasEnabledAtLeastOnce &&
+    immediateObject &&
+    typeof immediateObject === 'object' &&
+    INTERNALS in immediateObject
+  ) {
+    ;(immediateObject as NextImmediate)[Symbol.dispose]()
+  } else {
+    originalClearImmediate(immediateObject)
+  }
+}
+
+//========================================================
+
+const INTERNALS: unique symbol = Symbol.for('next.Immediate.internals')
+
+type NextImmediateInternals = {
+  /** Stored to reflect `ref()`/`unref()` calls, but has no effect otherwise */
+  hasRef: boolean
+  queueItem: ActiveQueueItem | null
+}
+
+/** Makes sure that we're implementing all the public `Immediate` methods */
+interface NativeImmediate extends NodeJS.Immediate {}
+
+/** Implements a shim for the native `Immediate` class returned by `setImmediate` */
+class NextImmediate implements NativeImmediate {
+  [INTERNALS]: NextImmediateInternals = {
+    queueItem: null,
+    hasRef: true,
+  }
+  hasRef() {
+    const internals = this[INTERNALS]
+    if (internals.queueItem) {
+      return internals.hasRef
+    } else {
+      // if we're no longer queued (cleared or executed), hasRef is always false
+      return false
+    }
+  }
+  ref() {
+    const internals = this[INTERNALS]
+    if (internals.queueItem) {
+      internals.hasRef = true
+    }
+    return this
+  }
+  unref() {
+    const internals = this[INTERNALS]
+    if (internals.queueItem) {
+      internals.hasRef = false
+    }
+    return this
+  }
+
+  /**
+   * Node invokes `_onImmediate` when an immediate is executed:
+   * https://github.com/nodejs/node/blob/42d363205715ffa5a4a6d90f4be1311487053d65/lib/internal/timers.js#L504
+   * It's visible on the public types, so we want to have it here for parity, but it's a noop.
+   * */
+  _onImmediate() {}
+
+  [Symbol.dispose]() {
+    // This is equivalent to `clearImmediate`.
+    const internals = this[INTERNALS]
+    if (internals.queueItem) {
+      // this is still queued. drop it.
+      const queueItem = internals.queueItem
+      internals.queueItem = null
+      clearQueueItem(queueItem)
+    }
+  }
+}
+
+// ==========================================
+
+// TODO: this causes the scheduler to loop -- apparently the write schedules a nextTick somewhere inside?
```

Comment:
maybe any immediate called while we're performing any of the administrative parts of the function needs to bypass the nextTick tracking?.